pkg/operator/controller/utils/test/events.go

```diff
@@ -0,0 +1,184 @@
+		EventType: we.Type,
+	}
+	switch obj := we.Object.(type) {
+	case *corev1.Secret:
+		te.ObjType = "secret"
+		te.Namespace = obj.Namespace
+		te.Name = obj.Name
+	case *corev1.ConfigMap:
+		te.ObjType = "configmap"
+		te.Namespace = obj.Namespace
+		te.Name = obj.Name
+	case *appsv1.Deployment:
+		te.ObjType = "deployment"
+		te.Namespace = obj.Namespace
+		te.Name = obj.Name
+	case *corev1.ServiceAccount:
+		te.ObjType = "serviceaccount"
+		te.Namespace = obj.Namespace
+		te.Name = obj.Name
+	case *rbacv1.ClusterRole:
+		te.ObjType = "clusterrole"
+		te.Name = obj.Name
+	case *rbacv1.ClusterRoleBinding:
+		te.ObjType = "clusterrolebinding"
+		te.Name = obj.Name
+	case *corev1.Namespace:
+		te.ObjType = "namespace"
+		te.Name = obj.Name
+	case *operatorv1alpha1.ExternalDNS:
+		te.ObjType = "externaldns"
+		te.Name = obj.Name
+	case *cco.CredentialsRequest:
+		te.ObjType = "credentialsrequest"
+		te.Name = obj.Name
+	}
+	return te
+}
+
+// Key returns a key like representation of the event.
+func (e Event) Key() string {
+	return string(e.EventType) + "/" + e.ObjType + "/" + e.Namespace + "/" + e.Name
+}
+
+// EventCollector collects all types of events for the given watch types.
+type EventCollector struct {
+	T          *testing.T
+	Client     client.WithWatch
+	WatchTypes []client.ObjectList
+	Verbose    bool
+	watches    []watch.Interface
+	eventsCh   chan watch.Event
+}
+
+// NewEventCollector returns an instance of the event collector.
+func NewEventCollector(t *testing.T, client client.WithWatch, watchTypes []client.ObjectList, bufSize int) *EventCollector {
+	return &EventCollector{
+		T:          t,
+		Client:     client,
+		WatchTypes: watchTypes,
+		eventsCh:   make(chan watch.Event, bufSize),
+	}
+}
+
+// Start starts watches for all the watch types.
+func (c *EventCollector) Start(ctx context.Context) {
+	c.T.Helper()
+
+	for _, watchType := range c.WatchTypes {
+		w, err := c.Client.Watch(ctx, watchType)
+		if err != nil {
+			c.T.Fatalf("failed to start the watch for %T: %v", watchType, err)
+		}
+		c.watches = append(c.watches, w)
+	}
+
+	// fan in the events
+	for _, w := range c.watches {
+		go func(ch <-chan watch.Event) {
+			for e := range ch {
+				if testing.Verbose() {
+					c.T.Logf("Got watch event: %v", e)
+				}
+				c.eventsCh <- e
+			}
+		}(w.ResultChan())
+	}
+}
+
+// Stop stops all the watches.
+func (c *EventCollector) Stop() {
+	for _, w := range c.watches {
+		w.Stop()
+	}
+}
+
+// Collect collects the given number of events.
+func (c *EventCollector) Collect(num int, timeout time.Duration) []Event {
+	return c.collectWithAction(num, timeout, func() {})
+}
+
+// CollectFatal collects the given number of events. Fails the test if the timeout is reached while not all events are received.
+func (c *EventCollector) CollectFatal(num int, timeout time.Duration) []Event {
+	c.T.Helper()
+
+	return c.collectWithAction(num, timeout, func() {
+		c.T.Fatalf("timed out waiting for all expected events")
+	})
+}
```

Comment:
This seems to be unused.

Reply:
yeah, I thought that maybe useful for someone to fail when the timeout is expired. But it's unpractical to keep all these functions if they are no use. Refactored back to 1 single function.